File: thornfield/trainer/core/triad.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple

import numpy as np

from .token import Token


@dataclass
class Triad:
    tokens: List[Token]
    joint_energy: float = 0.0
    convergence_delta: np.ndarray | None = None
    suggested_position: Tuple[int, int] | None = None

    def __post_init__(self) -> None:
        if len(self.tokens) != 3:
            raise ValueError("Triad must contain exactly 3 tokens")

    @property
    def is_class_diverse(self) -> bool:
        return len({t.token_class for t in self.tokens}) == 3

    @property
    def attractor_contribution(self) -> np.ndarray:
        weights = np.stack([t.attractor_weights for t in self.tokens], axis=0)
        return weights.mean(axis=0)

File: living_tales/ios/Models/Triad.swift
import Foundation

struct GridPosition: Hashable, Codable {
    let row: Int
    let col: Int
}

struct Triad: Identifiable, Hashable {
    let id: UUID
    let tokens: [Token]
    var jointEnergy: Float
    var convergenceDelta: [Float]
    var suggestedPosition: GridPosition

    init(tokens: [Token], jointEnergy: Float = 0.0, convergenceDelta: [Float] = [], suggestedPosition: GridPosition = GridPosition(row: 0, col: 0)) {
        self.id = UUID()
        self.tokens = tokens
        self.jointEnergy = jointEnergy
        self.convergenceDelta = convergenceDelta
        self.suggestedPosition = suggestedPosition
    }

    var isClassDiverse: Bool {
        Set(tokens.map(\.tokenClass)).count == 3
    }

    var attractorContribution: [Float] {
        let n = Float(tokens.count)
        guard let dims = tokens.first?.attractorWeights.count else { return [] }
        return (0..<dims).map { dim in
            tokens.reduce(0) { $0 + $1.attractorWeights[dim] } / n
        }
    }
}
